File: web/src/body/humanBodyModel.ts
import type { BodySide, SurfaceLocation } from '../models'

/**
 * Procedural, anatomically proportioned human body built from lofted elliptical
 * cross-sections. Body space: origin between the feet, +Y up, +Z toward the
 * front, +X toward the patient's LEFT (the body faces the viewer). Units: metres.
 *
 * Pure math — no Three.js dependency — so the same code drives geometry,
 * region mapping and tests.
 */

export const BodyPartId = {
  head: 'head', neck: 'neck', torso: 'torso',
  leftUpperArm: 'leftUpperArm', rightUpperArm: 'rightUpperArm',
  leftLowerArm: 'leftLowerArm', rightLowerArm: 'rightLowerArm',
  leftHand: 'leftHand', rightHand: 'rightHand',
  leftUpperLeg: 'leftUpperLeg', rightUpperLeg: 'rightUpperLeg',
  leftLowerLeg: 'leftLowerLeg', rightLowerLeg: 'rightLowerLeg',
  leftFoot: 'leftFoot', rightFoot: 'rightFoot',
} as const

export type Vec3 = [number, number, number]

export interface LoftSection {
  y: number
  centerX: number
  centerZ: number
  radiusX: number
  radiusZ: number
}

export interface LoftMesh {
  positions: Float32Array
  normals: Float32Array
  indices: Uint32Array
  /** Number of ring vertices (sections × segments); the two cap centres follow. */
  ringVertexCount: number
}

export class BodyPart {
  constructor(
    readonly id: string,
    readonly sections: LoftSection[],
    readonly segments: number = 40,
  ) {}

  get minY(): number { return this.sections[0].y }
  get maxY(): number { return this.sections[this.sections.length - 1].y }

  sectionAt(y: number): LoftSection {
    const s = this.sections
    if (y <= s[0].y) return s[0]
    if (y >= s[s.length - 1].y) return s[s.length - 1]
    for (let i = 1; i < s.length; i++) {
      if (y <= s[i].y) {
        const a = s[i - 1], b = s[i]
        const t = b.y - a.y > 0 ? (y - a.y) / (b.y - a.y) : 0
        return {
          y,
          centerX: a.centerX + (b.centerX - a.centerX) * t,
          centerZ: a.centerZ + (b.centerZ - a.centerZ) * t,
          radiusX: a.radiusX + (b.radiusX - a.radiusX) * t,
          radiusZ: a.radiusZ + (b.radiusZ - a.radiusZ) * t,
        }
      }
    }
    return s[s.length - 1]
  }

  /** Normalized height (0..1) and geometric angle in degrees (0 front, +90 patient's left). */
  surfaceParameters(p: Vec3): { height: number; angleDegrees: number } {
    const span = Math.max(this.maxY - this.minY, 1e-4)
    const height = (p[1] - this.minY) / span
    const s = this.sectionAt(p[1])
    const angleDegrees = (Math.atan2(p[0] - s.centerX, p[2] - s.centerZ) * 180) / Math.PI
    return { height, angleDegrees }
  }

  /** Exact inverse of `surfaceParameters` (accounts for elliptical sections). */
  surfacePoint(height: number, angleDegrees: number): { position: Vec3; normal: Vec3 } {
    const y = this.minY + (this.maxY - this.minY) * Math.min(Math.max(height, 0), 1)
    const s = this.sectionAt(y)
    const phi = (angleDegrees * Math.PI) / 180
    const rx = Math.max(s.radiusX, 1e-4), rz = Math.max(s.radiusZ, 1e-4)
    const theta = Math.atan2(rz * Math.sin(phi), rx * Math.cos(phi))
    const position: Vec3 = [s.centerX + rx * Math.sin(theta), y, s.centerZ + rz * Math.cos(theta)]
    const n: Vec3 = [Math.sin(theta) / rx, 0, Math.cos(theta) / rz]
    const len = Math.hypot(n[0], n[1], n[2])
    return { position, normal: [n[0] / len, n[1] / len, n[2] / len] }
  }
}

export function buildLoftMesh(part: BodyPart): LoftMesh {
  const m = part.segments
  const secs = part.sections
  const ringVertexCount = secs.length * m
  const vertexCount = ringVertexCount + 2
  const positions = new Float32Array(vertexCount * 3)
  let v = 0
  for (const s of secs) {
    for (let j = 0; j < m; j++) {
      const th = (j / m) * 2 * Math.PI
      positions[v++] = s.centerX + s.radiusX * Math.sin(th)
      positions[v++] = s.y
      positions[v++] = s.centerZ + s.radiusZ * Math.cos(th)
    }
  }
  const bottom = secs[0], top = secs[secs.length - 1]
  const bottomCenter = ringVertexCount, topCenter = ringVertexCount + 1
  positions.set([bottom.centerX, bottom.y, bottom.centerZ], bottomCenter * 3)
  positions.set([top.centerX, top.y, top.centerZ], topCenter * 3)

  const idx: number[] = []
  for (let si = 0; si < secs.length - 1; si++) {
    for (let j = 0; j < m; j++) {
      const jn = (j + 1) % m
      const a = si * m + j, b = si * m + jn, c = (si + 1) * m + j, d = (si + 1) * m + jn
      // Counter-clockwise when viewed from outside.
      idx.push(a, b, c, b, d, c)
    }
  }
  const base = (secs.length - 1) * m
  for (let j = 0; j < m; j++) {
    const jn = (j + 1) % m
    idx.push(bottomCenter, jn, j)
    idx.push(topCenter, base + j, base + jn)
  }
  const indices = Uint32Array.from(idx)

  const normals = new Float32Array(vertexCount * 3)
  for (let i = 0; i + 2 < indices.length; i += 3) {
    const ia = indices[i] * 3, ib = indices[i + 1] * 3, ic = indices[i + 2] * 3
    const abx = positions[ib] - positions[ia], aby = positions[ib + 1] - positions[ia + 1], abz = positions[ib + 2] - positions[ia + 2]
    const acx = positions[ic] - positions[ia], acy = positions[ic + 1] - positions[ia + 1], acz = positions[ic + 2] - positions[ia + 2]
    const nx = aby * acz - abz * acy, ny = abz * acx - abx * acz, nz = abx * acy - aby * acx
    for (const k of [ia, ib, ic]) { normals[k] += nx; normals[k + 1] += ny; normals[k + 2] += nz }
  }
  for (let i = 0; i < vertexCount; i++) {
    const o = i * 3
    const l = Math.hypot(normals[o], normals[o + 1], normals[o + 2])
    if (l > 0) { normals[o] /= l; normals[o + 1] /= l; normals[o + 2] /= l } else normals[o + 1] = 1
  }
  return { positions, normals, indices, ringVertexCount }
}

export class HumanBodyModel {
  constructor(readonly parts: BodyPart[], readonly selectablePartIds: ReadonlySet<string>) {}

  get height(): number { return Math.max(...this.parts.map((p) => p.maxY)) }

  part(id: string): BodyPart | undefined {
    return this.parts.find((p) => p.id === id)
  }

  surfaceLocation(partId: string, height: number, angleDegrees: number): SurfaceLocation | undefined {
    const part = this.part(partId)
    if (!part) return undefined
    const { position: [x, y, z], normal: [nx, ny, nz] } = part.surfacePoint(height, angleDegrees)
    return { meshId: partId, x, y, z, nx, ny, nz }
  }
}

// ---------------------------------------------------------------------------
// Default proportions (~1.75 m adult, neutral build).

function sec(y: number, centerX: number, centerZ: number, radiusX: number, radiusZ: number): LoftSection {
  return { y, centerX, centerZ, radiusX, radiusZ }
}

const torso = new BodyPart(BodyPartId.torso, [
  sec(0.790, 0, 0.000, 0.120, 0.080),
  sec(0.815, 0, 0.000, 0.160, 0.105),
  sec(0.860, 0, 0.000, 0.175, 0.118),
  sec(0.920, 0, 0.000, 0.172, 0.115),
  sec(0.980, 0, 0.000, 0.158, 0.105),
  sec(1.060, 0, 0.000, 0.148, 0.100),
  sec(1.150, 0, 0.002, 0.160, 0.108),
  sec(1.250, 0, 0.006, 0.176, 0.120),
  sec(1.340, 0, 0.002, 0.186, 0.112),
  sec(1.410, 0, 0.000, 0.192, 0.100),
  sec(1.445, 0, 0.000, 0.160, 0.085),
  sec(1.470, 0, 0.000, 0.105, 0.070),
  sec(1.490, 0, 0.000, 0.062, 0.056),
], 56)

const neck = new BodyPart(BodyPartId.neck, [
  sec(1.470, 0, -0.005, 0.058, 0.055),
  sec(1.560, 0, -0.005, 0.052, 0.050),
], 32)

function head(): BodyPart {
  const cy = 1.665, ry = 0.118, rx = 0.083, rz = 0.098, n = 12
  const sections: LoftSection[] = []
  for (let i = 0; i <= n; i++) {
    const phi = -Math.PI / 2 + (Math.PI * i) / n
    const scale = Math.max(Math.cos(phi), 0.02)
    const chin = phi < 0 ? 0.92 + 0.08 * (1 + Math.sin(phi)) : 1
    sections.push(sec(cy + ry * Math.sin(phi), 0, 0, rx * scale * chin, rz * scale))
  }
  return new BodyPart(BodyPartId.head, sections, 40)
}

function radiusAt(t: number, radii: [number, number][]): number {
  if (t <= radii[0][0]) return radii[0][1]
  const last = radii[radii.length - 1]
  if (t >= last[0]) return last[1]
  for (let i = 1; i < radii.length; i++) {
    if (t <= radii[i][0]) {
      const [ta, ra] = radii[i - 1], [tb, rb] = radii[i]
      return ra + (rb - ra) * ((t - ta) / Math.max(tb - ta, 1e-4))
    }
  }
  return last[1]
}

/** Tapered limb between two points with rounded ends. */
function limb(
  id: string, top: [number, number], bottom: [number, number],
  radii: [number, number][], opts: { z?: number; flattenZ?: number; segments?: number } = {},
): BodyPart {
  const { z = 0, flattenZ = 1, segments = 32 } = opts
  const steps = 14
  const sections: LoftSection[] = []
  for (let i = 0; i <= steps; i++) {
    const t = i / steps
    const x = bottom[0] + (top[0] - bottom[0]) * t
    const y = bottom[1] + (top[1] - bottom[1]) * t
    let r = radiusAt(t, radii)
    const edge = 0.08
    if (t < edge) r *= Math.sqrt(Math.max(0, 1 - ((edge - t) / edge) ** 2)) + 0.02
    if (t > 1 - edge) r *= Math.sqrt(Math.max(0, 1 - ((t - (1 - edge)) / edge) ** 2)) + 0.02
    sections.push(sec(y, x, z, r, r * flattenZ))
  }
  return new BodyPart(id, sections, segments)
}

function sideParts(side: BodySide): BodyPart[] {
  const sx = side === 'left' ? 1 : -1
  const L = side === 'left'
  return [
    limb(L ? BodyPartId.leftUpperArm : BodyPartId.rightUpperArm, [sx * 0.205, 1.425], [sx * 0.240, 1.120],
      [[0, 0.045], [0.3, 0.050], [0.75, 0.056], [1, 0.062]]),
    limb(L ? BodyPartId.leftLowerArm : BodyPartId.rightLowerArm, [sx * 0.242, 1.130], [sx * 0.268, 0.850],
      [[0, 0.032], [0.4, 0.040], [1, 0.048]]),
    limb(L ? BodyPartId.leftHand : BodyPartId.rightHand, [sx * 0.270, 0.855], [sx * 0.282, 0.680],
      [[0, 0.030], [0.5, 0.045], [1, 0.036]], { flattenZ: 0.45, segments: 24 }),
    limb(L ? BodyPartId.leftUpperLeg : BodyPartId.rightUpperLeg, [sx * 0.092, 0.880], [sx * 0.100, 0.470],
      [[0, 0.062], [0.15, 0.068], [0.6, 0.082], [1, 0.092]]),
    limb(L ? BodyPartId.leftLowerLeg : BodyPartId.rightLowerLeg, [sx * 0.100, 0.480], [sx * 0.100, 0.060],
      [[0, 0.038], [0.25, 0.048], [0.7, 0.062], [1, 0.058]], { z: -0.005 }),
    new BodyPart(L ? BodyPartId.leftFoot : BodyPartId.rightFoot, [
      sec(0.000, sx * 0.100, 0.045, 0.040, 0.115),
      sec(0.020, sx * 0.100, 0.045, 0.046, 0.125),
      sec(0.055, sx * 0.100, 0.025, 0.044, 0.095),
      sec(0.085, sx * 0.100, -0.005, 0.040, 0.060),
    ], 28),
  ]
}

export const standardBody = new HumanBodyModel(
  [torso, neck, head(), ...sideParts('left'), ...sideParts('right')],
  new Set([BodyPartId.torso, BodyPartId.leftUpperArm, BodyPartId.rightUpperArm, BodyPartId.leftUpperLeg, BodyPartId.rightUpperLeg]),
)

File: NeuproPatchTracker/Body/HumanBodyModel.swift
import Foundation
import simd

/// Identifiers of the body parts that make up the procedural human model.
enum BodyPartID {
    static let head = "head"
    static let neck = "neck"
    static let torso = "torso"
    static let leftUpperArm = "leftUpperArm"
    static let rightUpperArm = "rightUpperArm"
    static let leftLowerArm = "leftLowerArm"
    static let rightLowerArm = "rightLowerArm"
    static let leftHand = "leftHand"
    static let rightHand = "rightHand"
    static let leftUpperLeg = "leftUpperLeg"
    static let rightUpperLeg = "rightUpperLeg"
    static let leftLowerLeg = "leftLowerLeg"
    static let rightLowerLeg = "rightLowerLeg"
    static let leftFoot = "leftFoot"
    static let rightFoot = "rightFoot"
}

/// One elliptical cross-section of a lofted body part, in body space (metres).
struct LoftSection: Hashable, Sendable {
    var y: Float
    var centerX: Float
    var centerZ: Float
    var radiusX: Float
    var radiusZ: Float
}

/// A body part described as a stack of elliptical cross-sections.
///
/// Body space: origin between the feet, +Y up, +Z toward the front of the body,
/// +X toward the patient's LEFT (the body faces the viewer).
struct BodyPart: Identifiable, Hashable, Sendable {
    let id: String
    var sections: [LoftSection]
    var segments: Int = 40

    var minY: Float { sections.first?.y ?? 0 }
    var maxY: Float { sections.last?.y ?? 0 }

    /// Interpolated cross-section at height `y` (clamped to the part).
    func section(at y: Float) -> LoftSection {
        guard let first = sections.first, let last = sections.last else {
            return LoftSection(y: y, centerX: 0, centerZ: 0, radiusX: 0, radiusZ: 0)
        }
        if y <= first.y { return first }
        if y >= last.y { return last }
        for i in 1..<sections.count where y <= sections[i].y {
            let a = sections[i - 1], b = sections[i]
            let t = (b.y - a.y) > 0 ? (y - a.y) / (b.y - a.y) : 0
            return LoftSection(
                y: y,
                centerX: a.centerX + (b.centerX - a.centerX) * t,
                centerZ: a.centerZ + (b.centerZ - a.centerZ) * t,
                radiusX: a.radiusX + (b.radiusX - a.radiusX) * t,
                radiusZ: a.radiusZ + (b.radiusZ - a.radiusZ) * t)
        }
        return last
    }

    /// Normalized height (0 bottom ... 1 top) and angle in degrees around the
    /// part's axis (0 = front, +90 = patient's left, ±180 = back).
    func surfaceParameters(for point: SIMD3<Float>) -> (height: Float, angleDegrees: Float) {
        let span = max(maxY - minY, 0.0001)
        let height = (point.y - minY) / span
        let s = section(at: point.y)
        let angle = atan2(point.x - s.centerX, point.z - s.centerZ) * 180 / .pi
        return (height, angle)
    }

    /// Point and outward normal on the loft surface for given parameters.
    func surfacePoint(height: Float, angleDegrees: Float) -> (position: SIMD3<Float>, normal: SIMD3<Float>) {
        let y = minY + (maxY - minY) * min(max(height, 0), 1)
        let s = section(at: y)
        let phi = angleDegrees * .pi / 180
        let rx = max(s.radiusX, 0.0001), rz = max(s.radiusZ, 0.0001)
        // Convert the geometric angle (what `surfaceParameters` measures) to the
        // ellipse's parametric angle so the two functions are exact inverses.
        let theta = atan2(rz * sin(phi), rx * cos(phi))
        let position = SIMD3<Float>(
            s.centerX + rx * sin(theta), y, s.centerZ + rz * cos(theta))
        let normal = simd_normalize(SIMD3<Float>(sin(theta) / rx, 0, cos(theta) / rz))
        return (position, normal)
    }
}

/// Triangle mesh data for a lofted part (no RealityKit dependency).
struct LoftMesh: Sendable {
    var positions: [SIMD3<Float>]
    var normals: [SIMD3<Float>]
    var indices: [UInt32]

    /// Ring index (section) for each vertex; caps use the nearest section.
    var sectionIndex: [Int]
}

enum LoftMeshBuilder {
    static func build(_ part: BodyPart) -> LoftMesh {
        let m = part.segments
        var positions: [SIMD3<Float>] = []
        var sectionIndex: [Int] = []
        for (si, s) in part.sections.enumerated() {
            for j in 0..<m {
                let theta = Float(j) / Float(m) * 2 * .pi
                positions.append(SIMD3(
                    s.centerX + s.radiusX * sin(theta), s.y, s.centerZ + s.radiusZ * cos(theta)))
                sectionIndex.append(si)
            }
        }
        // Cap centres.
        let bottom = part.sections.first!
        let top = part.sections.last!
        let bottomCenter = UInt32(positions.count)
        positions.append(SIMD3(bottom.centerX, bottom.y, bottom.centerZ))
        sectionIndex.append(0)
        let topCenter = UInt32(positions.count)
        positions.append(SIMD3(top.centerX, top.y, top.centerZ))
        sectionIndex.append(part.sections.count - 1)

        var indices: [UInt32] = []
        for si in 0..<(part.sections.count - 1) {
            for j in 0..<m {
                let jn = (j + 1) % m
                let a = UInt32(si * m + j)
                let b = UInt32(si * m + jn)
                let c = UInt32((si + 1) * m + j)
                let d = UInt32((si + 1) * m + jn)
                // Counter-clockwise when viewed from outside.
                indices += [a, b, c]
                indices += [b, d, c]
            }
        }
        for j in 0..<m {
            let jn = (j + 1) % m
            indices += [bottomCenter, UInt32(jn), UInt32(j)]
            let base = (part.sections.count - 1) * m
            indices += [topCenter, UInt32(base + j), UInt32(base + jn)]
        }

        var normals = [SIMD3<Float>](repeating: .zero, count: positions.count)
        var i = 0
        while i + 2 < indices.count {
            let ia = Int(indices[i]), ib = Int(indices[i + 1]), ic = Int(indices[i + 2])
            let n = simd_cross(positions[ib] - positions[ia], positions[ic] - positions[ia])
            normals[ia] += n
            normals[ib] += n
            normals[ic] += n
            i += 3
        }
        normals = normals.map { simd_length($0) > 0 ? simd_normalize($0) : SIMD3(0, 1, 0) }
        return LoftMesh(positions: positions, normals: normals, indices: indices, sectionIndex: sectionIndex)
    }
}

/// The procedural, anatomically proportioned human body used by the app.
///
/// Every part is positioned in body space and rendered at the identity transform
/// under a single root entity, so `SurfaceLocation` local coordinates are body-space
/// coordinates. Replace this with a licensed USDZ by providing a different
/// `HumanBodyModel` (parts + parameterization) – see README "Body Model".
struct HumanBodyModel: Sendable {
    let parts: [BodyPart]
    /// Parts that may be hit-tested for placement (regions are defined on these).
    let selectablePartIDs: Set<String>

    /// Approximate overall height in metres.
    var height: Float { parts.map(\.maxY).max() ?? 1.75 }

    func part(withID id: String) -> BodyPart? {
        parts.first { $0.id == id }
    }

    func surfaceLocation(onPart partID: String, height: Float, angleDegrees: Float) -> SurfaceLocation? {
        guard let part = part(withID: partID) else { return nil }
        let p = part.surfacePoint(height: height, angleDegrees: angleDegrees)
        return SurfaceLocation(meshIdentifier: partID, position: p.position, normal: p.normal)
    }

    /// Body-space position for a stored location. Parts are at identity relative to the
    /// body root, so this is a direct read; kept as the single conversion point.
    func bodySpacePosition(for location: SurfaceLocation) -> SIMD3<Float> {
        location.position
    }

    static let standard = HumanBodyModel(parts: HumanBodyProportions.parts, selectablePartIDs: [
        BodyPartID.torso, BodyPartID.leftUpperArm, BodyPartID.rightUpperArm,
        BodyPartID.leftUpperLeg, BodyPartID.rightUpperLeg,
    ])
}

/// Default proportions (~1.75 m adult, neutral build).
enum HumanBodyProportions {
    static var parts: [BodyPart] {
        var result: [BodyPart] = [torso, neck, head]
        for side in BodySide.allCases {
            let sx: Float = side == .left ? 1 : -1
            result += [
                upperArm(side: side, sx: sx),
                lowerArm(side: side, sx: sx),
                hand(side: side, sx: sx),
                upperLeg(side: side, sx: sx),
                lowerLeg(side: side, sx: sx),
                foot(side: side, sx: sx),
            ]
        }
        return result
    }

    static let torso = BodyPart(id: BodyPartID.torso, sections: [
        LoftSection(y: 0.790, centerX: 0, centerZ: 0.000, radiusX: 0.120, radiusZ: 0.080),
        LoftSection(y: 0.815, centerX: 0, centerZ: 0.000, radiusX: 0.160, radiusZ: 0.105),
        LoftSection(y: 0.860, centerX: 0, centerZ: 0.000, radiusX: 0.175, radiusZ: 0.118),
        LoftSection(y: 0.920, centerX: 0, centerZ: 0.000, radiusX: 0.172, radiusZ: 0.115),
        LoftSection(y: 0.980, centerX: 0, centerZ: 0.000, radiusX: 0.158, radiusZ: 0.105),
        LoftSection(y: 1.060, centerX: 0, centerZ: 0.000, radiusX: 0.148, radiusZ: 0.100),
        LoftSection(y: 1.150, centerX: 0, centerZ: 0.002, radiusX: 0.160, radiusZ: 0.108),
        LoftSection(y: 1.250, centerX: 0, centerZ: 0.006, radiusX: 0.176, radiusZ: 0.120),
        LoftSection(y: 1.340, centerX: 0, centerZ: 0.002, radiusX: 0.186, radiusZ: 0.112),
        LoftSection(y: 1.410, centerX: 0, centerZ: 0.000, radiusX: 0.192, radiusZ: 0.100),
        LoftSection(y: 1.445, centerX: 0, centerZ: 0.000, radiusX: 0.160, radiusZ: 0.085),
        LoftSection(y: 1.470, centerX: 0, centerZ: 0.000, radiusX: 0.105, radiusZ: 0.070),
        LoftSection(y: 1.490, centerX: 0, centerZ: 0.000, radiusX: 0.062, radiusZ: 0.056),
    ], segments: 56)

    static let neck = BodyPart(id: BodyPartID.neck, sections: [
        LoftSection(y: 1.470, centerX: 0, centerZ: -0.005, radiusX: 0.058, radiusZ: 0.055),
        LoftSection(y: 1.560, centerX: 0, centerZ: -0.005, radiusX: 0.052, radiusZ: 0.050),
    ], segments: 32)

    static var head: BodyPart {
        let cy: Float = 1.665, ry: Float = 0.118, rx: Float = 0.083, rz: Float = 0.098
        var sections: [LoftSection] = []
        let n = 12
        for i in 0...n {
            let phi = -Float.pi / 2 + Float.pi * Float(i) / Float(n)
            let scale = max(cos(phi), 0.02)
            // Slightly narrower toward the chin.
            let chin: Float = phi < 0 ? 0.92 + 0.08 * (1 + sin(phi)) : 1
            sections.append(LoftSection(
                y: cy + ry * sin(phi), centerX: 0, centerZ: 0.0,
                radiusX: rx * scale * chin, radiusZ: rz * scale))
        }
        return BodyPart(id: BodyPartID.head, sections: sections, segments: 40)
    }

    /// Tapered limb between two points with rounded ends.
    private static func limb(
        id: String, from top: SIMD2<Float>, to bottom: SIMD2<Float>, z: Float = 0,
        radii: [(t: Float, r: Float)], flattenZ: Float = 1, segments: Int = 32
    ) -> BodyPart {
        var sections: [LoftSection] = []
        let steps = 14
        for i in 0...steps {
            let t = Float(i) / Float(steps)
            let x = bottom.x + (top.x - bottom.x) * t
            let y = bottom.y + (top.y - bottom.y) * t
            var r = radius(at: t, radii: radii)
            // Round the ends.
            let edge: Float = 0.08
            if t < edge { r *= sqrt(max(0, 1 - pow((edge - t) / edge, 2))) + 0.02 }
            if t > 1 - edge { r *= sqrt(max(0, 1 - pow((t - (1 - edge)) / edge, 2))) + 0.02 }
            sections.append(LoftSection(y: y, centerX: x, centerZ: z, radiusX: r, radiusZ: r * flattenZ))
        }
        return BodyPart(id: id, sections: sections, segments: segments)
    }

    private static func radius(at t: Float, radii: [(t: Float, r: Float)]) -> Float {
        guard let first = radii.first, let last = radii.last else { return 0.05 }
        if t <= first.t { return first.r }
        if t >= last.t { return last.r }
        for i in 1..<radii.count where t <= radii[i].t {
            let a = radii[i - 1], b = radii[i]
            let k = (t - a.t) / max(b.t - a.t, 0.0001)
            return a.r + (b.r - a.r) * k
        }
        return last.r
    }

    static func upperArm(side: BodySide, sx: Float) -> BodyPart {
        limb(id: side == .left ? BodyPartID.leftUpperArm : BodyPartID.rightUpperArm,
             from: SIMD2(sx * 0.205, 1.425), to: SIMD2(sx * 0.240, 1.120),
             radii: [(0, 0.045), (0.3, 0.050), (0.75, 0.056), (1.0, 0.062)])
    }

    static func lowerArm(side: BodySide, sx: Float) -> BodyPart {
        limb(id: side == .left ? BodyPartID.leftLowerArm : BodyPartID.rightLowerArm,
             from: SIMD2(sx * 0.242, 1.130), to: SIMD2(sx * 0.268, 0.850),
             radii: [(0, 0.032), (0.4, 0.040), (1.0, 0.048)])
    }

    static func hand(side: BodySide, sx: Float) -> BodyPart {
        limb(id: side == .left ? BodyPartID.leftHand : BodyPartID.rightHand,
             from: SIMD2(sx * 0.270, 0.855), to: SIMD2(sx * 0.282, 0.680),
             radii: [(0, 0.030), (0.5, 0.045), (1.0, 0.036)], flattenZ: 0.45, segments: 24)
    }

    static func upperLeg(side: BodySide, sx: Float) -> BodyPart {
        limb(id: side == .left ? BodyPartID.leftUpperLeg : BodyPartID.rightUpperLeg,
             from: SIMD2(sx * 0.092, 0.880), to: SIMD2(sx * 0.100, 0.470),
             radii: [(0, 0.062), (0.15, 0.068), (0.6, 0.082), (1.0, 0.092)])
    }

    static func lowerLeg(side: BodySide, sx: Float) -> BodyPart {
        limb(id: side == .left ? BodyPartID.leftLowerLeg : BodyPartID.rightLowerLeg,
             from: SIMD2(sx * 0.100, 0.480), to: SIMD2(sx * 0.100, 0.060), z: -0.005,
             radii: [(0, 0.038), (0.25, 0.048), (0.7, 0.062), (1.0, 0.058)])
    }

    static func foot(side: BodySide, sx: Float) -> BodyPart {
        // A short flattened loft; the foot extends forward via a shifted centre.
        let id = side == .left ? BodyPartID.leftFoot : BodyPartID.rightFoot
        return BodyPart(id: id, sections: [
            LoftSection(y: 0.000, centerX: sx * 0.100, centerZ: 0.045, radiusX: 0.040, radiusZ: 0.115),
            LoftSection(y: 0.020, centerX: sx * 0.100, centerZ: 0.045, radiusX: 0.046, radiusZ: 0.125),
            LoftSection(y: 0.055, centerX: sx * 0.100, centerZ: 0.025, radiusX: 0.044, radiusZ: 0.095),
            LoftSection(y: 0.085, centerX: sx * 0.100, centerZ: -0.005, radiusX: 0.040, radiusZ: 0.060),
        ], segments: 28)
    }
}
